lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetColumnChunkIterator.java

```diff
@@ -13,112 +13,110 @@
 import org.apache.parquet.column.Encoding;
-import org.apache.parquet.column.statistics.Statistics;
 import org.apache.parquet.format.DataPageHeader;
 import org.apache.parquet.format.DataPageHeaderV2;
 import org.apache.parquet.format.DictionaryPageHeader;
 import org.apache.parquet.format.PageHeader;
 import org.apache.parquet.format.Util;
 import org.apache.parquet.internal.column.columnindex.OffsetIndex;
 
+import javax.annotation.Nullable;
+
 import java.io.IOException;
-import java.util.LinkedList;
-import java.util.List;
 import java.util.Optional;
 import java.util.OptionalLong;
 
-import static com.google.common.base.Verify.verify;
 import static io.trino.parquet.ParquetTypeUtils.getParquetEncoding;
 import static java.util.Objects.requireNonNull;
+import static org.apache.parquet.format.PageType.DICTIONARY_PAGE;
 
-public class ParquetColumnChunk
+public class ParquetColumnChunkIterator
+        extends AbstractIterator<DataPage>
 {
     private final Optional<String> fileCreatedBy;
     private final ColumnChunkDescriptor descriptor;
-    private final List<Slice> slices;
-    private int sliceIndex;
-    private BasicSliceInput input;
+    private ChunkedInputStream input;
     private final OffsetIndex offsetIndex;
 
-    public ParquetColumnChunk(
+    // we read header of the first page to see if it is a dictionary page
+    private PageHeader firstPageHeader;
+    private DictionaryPage dictionaryPage;
+    private long valueCount;
+    private int dataPageCount;
+
+    public ParquetColumnChunkIterator(
             Optional<String> fileCreatedBy,
             ColumnChunkDescriptor descriptor,
-            List<Slice> slices,
+            ChunkedInputStream input,
             OffsetIndex offsetIndex)
+            throws IOException
     {
         this.fileCreatedBy = requireNonNull(fileCreatedBy, "fileCreatedBy is null");
-        this.descriptor = descriptor;
-        this.slices = slices;
-        this.sliceIndex = 0;
-        this.input = slices.get(0).getInput();
+        this.descriptor = requireNonNull(descriptor, "descriptor is null");
+        this.input = requireNonNull(input, "input is null");
         this.offsetIndex = offsetIndex;
-    }
-
-    private void advanceIfNecessary()
-    {
-        if (input.available() == 0) {
-            sliceIndex++;
-            if (sliceIndex < slices.size()) {
-                input = slices.get(sliceIndex).getInput();
+        if (hasMorePages(valueCount, dataPageCount)) {
+            PageHeader pageHeader = readPageHeader();
+            if (pageHeader.type.equals(DICTIONARY_PAGE)) {
+                dictionaryPage = readDictionaryPage(pageHeader, pageHeader.getUncompressed_page_size(), pageHeader.getCompressed_page_size());
+            }
+            else {
+                firstPageHeader = pageHeader;
             }
         }
     }
 
-    protected PageHeader readPageHeader()
-            throws IOException
+    @Nullable
```

Comment:
I see that it should be `@Nullable`.
Do we gain anything by extending from `AbstactIterator` ?
Implementing `Iterator` interface might be more natural here since we have `hasMorePages` method already

Reply:
let me try implementing `Iterator` directly

Reply:
done.